### ISSUE TITLE
0.4.* unavailable after shower turns on

### ISSUE THREAD

**Issue by 38Cherry** [niemyjski/homeassistant-kohler#35]
[home-assistant_kohler_2024-11-27T15-10-06.061Z.log](https://github.com/user-attachments/files/17936473/home-assistant_kohler_2024-11-27T15-10-06.061Z.log)

Since 0.4.x came out when i start my shower the controller becomes unavailable and generally doesn't come back online without a power cycle to the Kohler controller.  It is not pingable, reachable via http, and the Kohler Konnect app is not working.  Attached is a debug file from HA.  First thing i did was update to 0.4.1 then started the shower with HA.  As it went through the warm up cycle the status was correct and valves updated correctly.  Here's where i wasn't sure what triggered everything to go unavailable.  It was either as it moved from warm up to operation or when I touched the touchscreen to turn on the screen.

**Comment by niemyjski** [comment]
Thanks for this, I was able to experience this myself on 0.4.0-pre but not sense. If you could get exact reproduction steps that would be great. I'll also try it as well. I had it happen to me when I manually turned on valves in rapid succession via HA. 

@dcmeglio I wonder if it could be something with how we call the api to get values. Like maybe the coordinator is going wide on getting values with x requests at the same time?

**Comment by 38Cherry** [comment]
I didn't reboot the controller this morning after it became unavailable.  It went unavailable at 950am ish and just leaving it alone it came back online at 1236pm.

Seems like the controller is getting spammed by ha and locks it up for a couple hours and eventually eased up and recovers.  Unfortunately I turned off debugging earlier

**Comment by 38Cherry** [comment]
I was able to recreate it just now and i was very deliberate in watching what i did.  I started the shower with my HA automation (pressing my insteon button).  shower warmed up (HA didn't change status until warm up was complete.  This is the same behavior as the Kohler Konnect App, which sends the shower ready notification at the same time.   It turned on with a strange sequence of valves that i never turn on (valve 1,2 and 3).  I turned off 3 using the switch in HA.  It turned off fine.  Waited about 10 seconds then turned off valve 2 and that immediately caused the controller to go unavailable.  

One thing to note is that HA is not going unavailable, it is the actual Kohler controller because all comms are locked up including ping and browser.  Not sure if i made that clear before.  I'll leave it alone so it comes back again and see if i can pull the debug logs from the controller.



[home-assistant_kohler_2024-11-27T17-47-59.021Z.log](https://github.com/user-attachments/files/17938271/home-assistant_kohler_2024-11-27T17-47-59.021Z.log)

**Comment by dcmeglio** [comment]
I haven't experienced this issue myself with the new version. But I'll say I have experienced it in the past. If you poll the controller too frequently, it can sometimes lock up. Probably a cruddy HTTP server it's running. I believe the previous integration used the default polling interval of 30s, this now has the update_interval set to 20s. Maybe that's too fast?

**Comment by niemyjski** [comment]
Yeah, maybe we should back off the frequency since we have the more real time polling once we know it's active.

**Comment by dcmeglio** [comment]
Just wanted to call out, I haven't had this happen to me. Is it still occurring?

**Comment by 38Cherry** [comment]
I downgraded to 0.3.2 when 0.4.x came out because it was frequent.  I can
upgrade again and let you know.

**Comment by dcmeglio** [comment]
That would be helpful. I've been running it for weeks and had no issues.

**Comment by niemyjski** [comment]
I have not seen it but I also haven't been controlling valves via ha. Would be good to denounce updates?

**Comment by 38Cherry** [comment]
I went back to 4.2 and first shower I went into them unavailable again.
Maybe it's something in my setup but logs don't really show anything

**Comment by dcmeglio** [comment]
And the controller locks up?

**Comment by 38Cherry** [comment]
Only the web interface/api.   The touch pad still works.  Iirc Kohler
connect stops working too.

If I leave it for a few hours it will come back up life.  It seems like the
controller is just looping or something.  It may be my controller has an
issue, idk.

**Comment by dcmeglio** [comment]
I have definitely experienced it if I poll to fast. For mine though I discovered the limit is 14 seconds. The integration is currently configured to poll every 20. We could try adjusting it back to the 30s it used to be and see if that helps you

**Comment by niemyjski** [comment]
I think the problem could be dosing the controller via ha scheduled updates when we preform actions quickly. Would be good to look into this and maybe make the interval configurable.

**Comment by 38Cherry** [comment]
i upgraded HA to 2025.1.1 today and it looks like 0.3.2 will not run now.  I upgraded to 0.4.2 and it ran but as soon as someone took a shower the controller locked up again.  Not sure  if it's my controller or what but i'm basically stuck now between HA and Kohler updates

**Comment by niemyjski** [comment]
Try power cycling the controller.

**Comment by 38Cherry** [comment]
still having issues here.  i have to stay on 0.3.2 to avoid the crashing of my controller but HA 2025.1.x will not run 0.3.2  as there is some library issue.  I haven't had time to dig inot the errors or branch out 0.3.2 to fix the breaking code.  for now i'm stuck with 2024.12.x

**Comment by dcmeglio** [comment]
When your controller logs up, are you starting the shower from HA? Or just starting the shower normally? If you're starting it from HA, https://github.com/niemyjski/homeassistant-kohler/pull/41 *might* fix it, or at least help, but can't say for sure since I haven't reproduced your problem myself yet.

**Comment by niemyjski** [comment]
@dcmeglio thanks a bunch it looks great. @38Cherry can you please try this out via https://github.com/niemyjski/homeassistant-kohler/releases/tag/0.4.3

**Comment by 38Cherry** [comment]
no just from normal operation of the shower.  usually when turning on or
off valves via the touchscreen.

**Comment by 38Cherry** [comment]
Thank you.  This did resolve the issue with HA 2025.1 buti will have to let
you know if it fixes the controller lock up issue.  If not we may need to
put the depreciation fix into a 0.3.x update.

thanks
Steve

**Comment by 38Cherry** [comment]
Controller lock-up still occurs.  Only locks up when shower is on and usually when changing state of valves using either the app but most frequently we use the touch panel.  Perhaps the best solution is to release the depreciation update without the thermostat/valves update as 0.3.3

**Comment by niemyjski** [comment]
I don't really have any intention of backporting or maintaining multiple releases as I have very limited resources and am grateful for everyones help especially @dcmeglio . If someone wants to take that lead that would be good but I think it would be even better to find the core issue and let's focus on that.

**Comment by niemyjski** [comment]
With that said, I haven't seen any lockups on my side but I'm not turning on valves, but I am using ha to turn on the shower.

**Comment by dcmeglio** [comment]
I agree, not to downplay the issue -- I want to fix it, but we know of a single person having the issue. So no evidence there's some super widespread bug. We are working on this in our spare time, so having to maintain two versions gets complicated and I think is a step backwards. Are you comfortable modifying a python file slightly? Since I can't reproduce your issue I can't really test here. It feels like the polling interval needs to be dialed in.

I suspect this needs to be increased for you, maybe 30 seconds? https://github.com/niemyjski/homeassistant-kohler/blob/master/custom_components/kohler/__init__.py#L212

If you are comfortable SSHing into your HA, you can open /config/custom_components/kohler/__init__.py and change that line. Then restart HA and see if it helps. If 30 secs works, what about 25? If 30 doesn't work, what about 35? It's going to be a bit of trial and error.

**Comment by mobrien118** [comment]
I would like to add to this bug. I just got the integration set up yesterday and this morning went to test.

When I try to activate a valve from the HA interface, not only does the system immediately become unavailable, and the valves do *not* start, my valves disappear from the touchscreen UI altogether and don't show back up until I power cycle the controller.

Happy to help troubleshoot in any way I can. I have a good deal of programming experience but not in the HA space. Happy to learn whatever is needed, but in the near term if there is anything I can log or diagnose, let me know.

**Comment by 38Cherry** [comment]
I ended up fixing version 0.3.2 for me to use.  I can probably create a pr
if we want to update a branch.  It's only three lines of code.

**Comment by niemyjski** [comment]
Might be helpful for you to share your firmware versions and how many ports you have. I have 4 ports and on latest firmware (I think I already posted in this thread), We run the shower at least twice a day and I haven't noticed this issue in months.

**Comment by mobrien118** [comment]
Just so I'm clear, what are you asking for when you ask to add some logging?

I have a 2x2 setup.
![Image](https://github.com/user-attachments/assets/67cef9e2-b582-418f-8740-50fe5dfb51e7)

Worth noting that the device "becomes unavailable" for a few seconds every couple hours or so, but immediately comes back. This could be due to the fact I'm running it across a wifi bridge. However, it only when, and every single time that, I try to turn on the shower from HA that it completely bombs out and removes the valves from the touchsereen, requiring a reboot.

My firmware is 0.0.3.89 and I have scoured the internet for any place to find out whether that is current or to find a download for anything newer and have failed. When I moved in here and first connected it to the internet, it did have some updates, though, so I assume it is as current as Kohler wants it to be...

**Comment by dcmeglio** [comment]
Any chance you can try what I suggested here? https://github.com/niemyjski/homeassistant-kohler/issues/35#issuecomment-2641403832

I also don't know if that's the latest version because they don't publish info, but it's the same version I'm on.

**Comment by 38Cherry** [comment]
Mobeien's issue is very similar but I never had my control panel freeze
up.  Everything else mentioned is the same.  Mine did come back after about
3 hours.

I ran out of wife points troubleshooting and had to just get it fixed.
Sorry I do want it to work, but version 4 didn't provide anything that I
saw as needed.

**Comment by mobrien118** [comment]
First off -  LOL @ "wife points" - I know exactly what you are talking about and I never had a term for it, until now!

Next order of business, I don't have to SSH in because I have the "Studio Code Server" addon installed (*HIGHLY RECOMMEND*) - I made the change and will test some time this weekend (when my wife is not around so I can save up wife points!)

Lastly, just to clarify, the touchscreen doesn't freeze up, the shower valves just completely disappear from it. In fact, the button "Start" or "Shower" (I am not at home, but it is the only button in the second row up and, I think it is green). When that button disappears, the touchscreen continues to function, otherwise, and, from my memory, I can somehow get to the valves screens, but they are then empty. I have not attempted to HTTP into the web interface when it is in the state, but I had assumed that it becoming "unavailable" in HA meant the web endpoint was dead.

Obviously, their system was only intended to be called from their web-UI as it should not be possible for a remote call to diminish or disable functionality unexpectedly - poor cost-cutting design practice if you ask me, but I understand companies' need to get things done cheaply, so ¯\_(ツ)_/¯

**Comment by dcmeglio** [comment]
`I have not attempted to HTTP into the web interface when it is in the state, but I had assumed that it becoming "unavailable" in HA meant the web endpoint was dead`
It could be, or it could just be a bug in our HA integration. Showing "unavailable" in HA doesn't automatically mean the device is dead, it's kind of a generic state in HA that means "some unhandled error happened" so it's hard to know. If you have logs when it's frozen up that would be super helpful in addition to the change I suggested.

`Obviously, their system was only intended to be called from their web-UI as it should not be possible for a remote call to diminish or disable functionality unexpectedly`
Unfortunately, the problem is you have an amazing bathroom hardware company like Kohler getting into the tech business. It's not their expertise. They outsourced building this device, but then you have people cutting corners and doing stupid things. What I've observed is the Kohler DTV allows very low rates of requests otherwise it gets overwhelmed. Think of it as a DoS attack on a website. Unfortunately whereas that usually requires millions of users hitting the website uncontrollably, the Kohler DTV seems to crumble even if I open the web interface in 5 tabs at once!

**Comment by niemyjski** [comment]
I wanted to update this issue a few weeks ago but was unable to. I ran into this issue once while using the controls. I feel like we need to add debug logging around any call site that schedules a data update or anything that would talk to the controller via the api. I think this is needed so we can see if we are DOSing the controller by many scheduled updates. Would anyone be able to do this?

**Comment by mobrien118** [comment]
Sorry, I took a video of it happening. It seems like there is possibly a component of both the "web interface" being overwhelmed as well as the touchscreen interface (assuming the touchscreen itself is making calls to an interface of its own and has its own processing resources for the GUI itself). I either mischaracterized/misrepresented the issue before, or it is manifesting slightly differently now.

https://github.com/user-attachments/assets/6e1b8c16-6a6a-4769-9386-d6961437972c

One idea, if we think spamming is overloading the controller's available resources, would be to treat the interface as a bus type and create a call queue where triggering an action simply queues it and the same number of calls is made per second regardless of what HA wants to do. Checking controller status would be an "idle state" call only. This might create small perceivable delays in the HA interface to the end user, but may be necessary since, yes, the Kohler hardware and software are not designed for our particular use case. Unfortunately, I'm also not familiar with HA very much to know whether this is possible or practical - just an idea.

It might be helpful to know (scientifically) what types of limitations the system actually has regarding calls/second, as well as how varied different hardware/software/(environmental conditions) affect the limitations and tailor this module to the least common denominator. In my experience, if they come out with any more software updates, they are most likely to only further bloat the system and further stretch the available resources. It is improbable that they will intentionally introduce efficiency and stability, since they probably have no incentive to do so.

**Comment by niemyjski** [comment]
I dug into this again on March 18, 2026 after seeing a report that opening one outlet while the shower was off could start multiple outlets.

A useful finding from the controller's own UI is that it treats outlet selection while the shower is off differently from live outlet changes while running:
- while off, outlet selection is just a remembered/default selection set
- while running, outlet selection changes send updates immediately

The integration was treating those as the same thing.

I landed a fix locally that:
- serializes API reads and writes through one lock
- debounces/coalesces rapid `quick_shower` changes so only the latest desired state is sent
- replaces immediate post-command refresh spam with one delayed refresh
- models off-state selected/default outlets separately from live running outlets
- makes `open` on one outlet while off start only that requested outlet instead of inheriting stale multi-outlet state
- makes shower start use the controller's default/selected outlet state instead of inferring from currently open outlets

Local verification passed:
- `./.venv/bin/ruff check .`
- `./.venv/bin/ruff format --check .`
- `./.venv/bin/pytest` (`18 passed`)

I have not validated active user preset behavior end-to-end yet. I split that into separate follow-up work because it may be a distinct controller flow from the outlet-start bug.

**Comment by niemyjski** [comment]
https://github.com/niemyjski/homeassistant-kohler/releases/tag/0.5.0

**Comment by 38Cherry** [comment]
Just to close the loop from way back.  I updated to this new version and have had none of the disconnect problems that i had with 0.4.*  .  I'd also say that the issue where the shower would turn off after the pre-heat when started with HA has greatly reduced.  So overall great job team.
